Loading Screen

Triangle345 opened the pull request:
Loading scene and revamp of multi threading loading functionality.

Triangle345 committed:
Merge branch 'master' of https://github.com/FieldWarning/projectFieldWarning

Triangle345 committed:
- Changed text to better quality in loading screen
- TerrainMap will now write the entire map to the compressed file

Triangle345 committed:
Merge branch 'master' of https://github.com/FieldWarning/projectFieldWarning

Triangle345 committed:
Revamp of loading screen/multithreading part 1

Triangle345 committed:
worker threads mostly working now with scene transition

Triangle345 committed:
Big commit here..
- All full scenes now use the loading-scene to populate data assets like terrain and pathfinding. The LoadingScreen in this scene then loads the destinationScene with the loaded objects as a DoNotDestroyOnLoad Prefab/LoadedData.prefab

- Redesigned the the way we use multithreading to load various data. We now have a choice to use Coroutines for data that is required to load in the main thread. Coroutines are a lot slower so its recommended to use regular multithreading if possible.

- Created a loading folder under assets to organize the new (and old) multithreading loader screen files.

Triangle345 committed:
removed pathfinder binary on disk data.. this should be getting generated everytime its not there.. it sucks.. but eventually we need a build system that will produce this for us with the released product

Triangle345 committed:
Pushed files that didnt get added

Triangle345 committed:
added loader file which was missing from master

Triangle345 committed:
added more missing files

pvutov commented on src/FieldWarning/Assets/Terrain/Scripts/TerrainMap.cs:
```diff
@@ -43,32 +45,39 @@ public class TerrainMap
     public readonly Vector3 MapMin, MapMax, MapCenter;
 
+    // internal variable used for main thread notification
+    private bool _processing = false;
```
generates a warning about unused variable

Triangle345 replied:
will remove

pvutov commented on src/FieldWarning/Assets/Terrain/Scripts/TerrainMap.cs:
```diff
@@ -43,32 +45,39 @@ public class TerrainMap
     List<Vector3> _bridgePositions = new List<Vector3>();
 
     // this is only needed for map testing
     private byte[,] originalTestMap = null;
```
generates a warning about unused variable

Triangle345 replied:
hmm this one is definitely used. I'm not saying any unused warning for that.

pvutov commented on src/FieldWarning/Assets/Loading/CoroutineRunner.cs:
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class CoroutineRunner : MonoBehaviour
```
License + namespace + comment (something like "We need a monobehavior to start coroutines on so we use this empty one"), also no need to have the start and update methods.

pvutov commented on src/FieldWarning/Assets/Model/Game/MatchSession.cs:
```diff
@@ -131,6 +147,18 @@ public void Awake()
+                SceneManager.LoadScene(4, LoadSceneMode.Single);
+            }
+
+            
```
private void Start + less empty newlines

pvutov commented on src/FieldWarning/Assets/Model/Game/MatchSession.cs:
```diff
@@ -131,6 +147,18 @@ public void Awake()
+            if (_loadedData == null)
+            {
+                LoadingScreen.destinationScene = SceneManager.GetActiveScene().buildIndex;
+                SceneManager.LoadScene(4, LoadSceneMode.Single);
```
loading a scene by index like this is error-prone since we will change the scene ordering and count around, you can also load by name

pvutov commented on src/FieldWarning/Assets/Terrain/Scripts/TerrainMap.cs:
```diff
@@ -129,94 +138,109 @@ public TerrainMap(Terrain[] terrains1D)
     }
 
-    private void MapTester()
+    private string getMapCachePath()
```
GetMapCachePath, ideally 'GetHeightMapCachePath'

Triangle345 replied:
this is not really a path for height maps.. just general terrain maps.. Currently we dont generate a height map, only a terrain map

pvutov replied:
GetTerrainMapCachePath then :)

pvutov commented on src/FieldWarning/Assets/Util/DontDestroyOnLoad.cs:
```diff
@@ -0,0 +1,16 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class DontDestroyOnLoad : MonoBehaviour
```
License + namespace + comment on the purpose of this class would be nice (its obvious in the context of this pull request, but a bit strange if you're just browsing the code base and happen to open it).

pvutov commented on src/FieldWarning/Assets/Util/DontDestroyOnLoad.cs:
```diff
@@ -0,0 +1,16 @@
+    public int id = 0;
+
+    // Start is called before the first frame update
+    void Awake()
```
private void Awake

+ if id is needed, it should be called 'Id'. But I think you only increment it just before destroying the object, never reading it?

Triangle345 replied:
yep its there to determine if duplicates were saved. It helps determine what objects to remove

pvutov commented on src/FieldWarning/Assets/Loading/Loader.cs:
```diff
@@ -0,0 +1,90 @@
+            _currentWorker.PercentDone = percent;
+    }
+
+    public bool isFinished()
```
IsFinished

pvutov commented on src/FieldWarning/Assets/Loading/LoadingScreen.cs:
```diff
@@ -12,51 +12,56 @@
+    public GameObject loadedData;
+
+
+    public void Awake()
```
private

pvutov commented on src/FieldWarning/Assets/Loading/Worker.cs:
```diff
@@ -0,0 +1,102 @@
+    void Update()
+    {
+
+    }
```
remove empty methods + add license and ideally namespace

pvutov commented on src/FieldWarning/Assets/Loading/Worker.cs:
```diff
@@ -0,0 +1,102 @@
+        PercentDone = 0;
+    }
+
+    protected void Ended()
```
These aren't really function names. I agree that it's hard to name them but with the current names I had to look at the implementation to get what they're doing.

"SetProgressToStarted" and "SetProgressToFinished'?

pvutov commented on src/FieldWarning/Assets/Units/Component/Movement/PathfinderData.cs:
```diff
@@ -49,12 +53,14 @@ public PathfinderData(TerrainMap map)
             string scenePathWithFilename = SceneManager.GetActiveScene().path;
             string sceneDirectory = Path.GetDirectoryName(scenePathWithFilename);
-            string graphFile = Path.Combine(sceneDirectory, sceneName + GRAPH_FILE_SUFFIX);
+            _graphFile = Path.Combine(sceneDirectory, sceneName + map.TerrainId + GRAPH_FILE_SUFFIX);
```
There is a .gitignore file in Scene/ iirc, you should add the new generated files to it. This should do the job:
*_map_terrain_cache.dat
 *_map_terrain_cache.dat.meta
 *_pathfinder_graph.dat
 *_pathfinder_graph.dat.meta

pvutov commented on src/FieldWarning/Assets/Model/Game/MatchSession.cs:
```diff
@@ -131,6 +147,18 @@ public void Awake()
+            {
+                LoadingScreen.destinationScene = SceneManager.GetActiveScene().buildIndex;
+                SceneManager.LoadScene(4, LoadSceneMode.Single);
+            }
```
else 
            {
#if UNITY_EDITOR
                // Default to hosting if entering play mode directly into a match scene:
                if (!NetworkClient.isConnected)
                    _networkManager.StartHost();
#endif
            }


+ remove the similar fragment from Awake()

Triangle345 committed:
Resolved issues from code review

pvutov commented on src/FieldWarning/Assets/Loading/CoroutineRunner.cs:
```diff
@@ -0,0 +1,29 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Loading
```
All of our namespaces start with PFW (in this case it would be PFW.Loading). Also, would be nice to add it to the other files in the loading folder too.

pvutov commented on src/FieldWarning/Assets/Model/Game/MatchSession.cs:
```diff
@@ -76,8 +77,11 @@ public class MatchSession : MonoBehaviour
+
         public void Awake()
         {
+            Debug.Log("Match Session started.");
```
Do you still need this debug message?

pvutov commented on src/FieldWarning/Assets/Model/Game/MatchSession.cs:
```diff
@@ -117,18 +121,33 @@ public void Awake()
+            {
+                LoadingScreen.destinationScene = SceneManager.GetActiveScene().buildIndex;
+                SceneManager.LoadScene("loading-scene", LoadSceneMode.Single);
+            } else
```
}
else 
{

pvutov commented on src/FieldWarning/Assets/Camera/MiniMap.cs:
```diff
@@ -50,7 +50,9 @@ private void Start()
         TerrainMap map = MatchSession.Current.TerrainMap;
-
+        if (map == null)
+            return;
```
For these random null checks we should write a comment on why we added them. Something like

> // When loading a map, the camera is moved to the loading scene,
> // where the terrain map is not yet loaded. Don't throw errors in that case.
> if (map == null)
>     return;

Otherwise this code is very ??? when you encounter it.

Also, I am still getting an error in Assets/Camera/SlidingCameraBehaviour.cs:568 (MaybeChangeTerrainMaterial())

pvutov commented on src/FieldWarning/Assets/Camera/SlidingCameraBehaviour.cs:
```diff
@@ -354,6 +354,12 @@ private void TiltCameraIfNearGround(float oldAltitude)
+        {
+            Debug.Log("Unable to clamp camera altitude. No terrain data available.");
+            return;
+        }
```
Same as the minimap remark, needs a comment, and probably should not log anything.

pvutov commented on src/FieldWarning/Assets/Camera/SlidingCameraBehaviour.cs:
```diff
@@ -478,6 +484,13 @@ private ScreenCorner GetScreenCornerForMousePosition(Vector2 mousePosition)
+        {
+            Debug.Log("Unable to clamp camera to XZ position. No map data loaded.");
+            return;
+        }
```
Same as the minimap remark, needs a comment, and probably should not log anything.

pvutov commented on src/FieldWarning/Assets/Loading/LoadedData.cs:
```diff
@@ -0,0 +1,45 @@
+    void Update()
+    {
+
+    }
```
Remove update and make Start private.

pvutov commented on src/FieldWarning/Assets/Terrain/Scripts/TerrainMap.cs:
```diff
@@ -20,8 +20,10 @@
+//using static Loading;
 
-public class TerrainMap
+public class TerrainMap :Loader
```
: Loader

Triangle345 committed:
added loading.meta file for the loading folder

Triangle345 committed:
- Fixed a few error messages before the terrain data is loaded.
Now object gets destroyed on DontDestroyOnLoad... It doesnt destroy on load but DontDestroyOnLoad now chooses to destroy itself if its a duplicate.

- Added a place holder for terrain id in TerrainDebug.. maybe someday it would be useful to see what terrain maps to which world space.
